src/needless_continue.rs

```diff
@@ -0,0 +1,547 @@
+//! Checks for continue statements in loops that are redundant.
+//!
+//! For example, the lint would catch
+//!
+//! ``` 
+//! while condition() {
+//!     update_condition();
+//!     if x {
+//!         // ...
+//!     } else {
+//!         continue;
+//!     }
+//!     println!("Hello, world");
+//! }
+//! ```
+//!
+//! And suggest something like this:
+//! 
+//! ```
+//! while condition() {
+//!     update_condition();
+//!     if x {
+//!         // ...
+//!         println!("Hello, world");
+//!     }
+//! }
+//!
+//! This lint is **warn** by default.
+
+use std;
+use rustc_front::print::pprust;
+use rustc::lint::*;
+use rustc_front::hir::*;
+use syntax::codemap::{Span,Spanned,BytePos,Pos,ExpnId};
+use syntax::ptr::P;
+use utils::{in_macro, span_help_and_lint, snippet};
+use self::LintType::*;
+
+const DROP_ELSE_BLOCK_AND_MERGE_MSG: &'static str =
+    "Consider dropping the else clause and merging the code that follows (in the loop) with the if block, like so:\n";
+
+const DROP_ELSE_BLOCK_MSG: &'static str =
+    "Consider dropping the else clause, which leaves the if expression as follows:\n"; 
+
+const REWRITE_IF_COND_MSG: &'static str = "Consider rewriting that code as follows:\n";
+
+declare_lint! {
+    pub NEEDLESS_CONTINUE,
+    Warn,
+    "`continue` statements that can be replaced by a rearrangement of code"
+}
+
+#[derive(Copy,Clone)]
+pub struct NeedlessContinue;
+
+impl LintPass for NeedlessContinue {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(NEEDLESS_CONTINUE)
+    }
+}
+
+impl LateLintPass for NeedlessContinue {
+    fn check_expr(&mut self, ctx: &LateContext, expr: &Expr) {
+        if !in_macro(ctx, expr.span) {
+            check_and_warn(ctx, expr);
+        }
+    }
+}
+
+fn check_and_warn(ctx: &LateContext, expr: &Expr) {
+    let mut found = false;
```

Comment:
I think we could use an inner function instead of this mutable variable, but I don't care too much either way.